### PR TITLE
Add pwm control for Aquaero

### PR DESCRIPTION
**Pull request by leoratte**
Added option to control the fans with static pwm. The control works in two steps: First the pwm value is written to preset controller. For simplicity I use the first controller for the first fan and so on. Second the ID of the preset controller is used as the fan control source.

### CODE REVIEW

File: aquacomputer_d5next.c [hunk 1]
```diff
@@ -19,6 +19,7 @@
 #include <linux/mutex.h>
 #include <linux/seq_file.h>
 #include <asm/unaligned.h>
+#include <linux/delay.h>
```
Comment (aleksamagicka) [inline aquacomputer_d5next.c:22]: Includes must be sorted, with the `asm` includes at the end (not sure why that is so, but that's what I've been told on the LKML.)

File: aquacomputer_d5next.c [hunk 2]
```diff
@@ -652,12 +668,20 @@ static int aqc_read(struct device *dev, enum hwmon_sensor_types type, u32 attr,
 			*val = *val + 1;	/* Incrementing to satisfy hwmon rules */
 			break;
 		case hwmon_pwm_input:
-			ret =
-			    aqc_get_ctrl_val(priv,
+			if (priv->kind == aquaero) {
+				ret =
+					aqc_get_ctrl_val(priv,
+					    AQUAERO_CTRL_PRESET_OFFSET + channel * 2, val, 16);
```
Comment (aleksamagicka) [inline aquacomputer_d5next.c:674]: Extract the `2` into a macro like `AQUAERO_CTRL_PRESET_SIZE`, similar to the `AQC_TEMP_SENSOR_SIZE`, so we don't have unlabeled magic constants.

**Review by aleksamagicka — CHANGES_REQUESTED** [review]
Just minor comments here. I've tested it, works for PWM values bigger than 100. That's probably due to some minimum RPM setting, I'll check in aquasuite.

**Comment by leoratte** [comment]
The PWM values range from 0 to 255. If you want to check the current configuration in aquasuite you may have to remove the device and add it again in aquasuite to show the correct configuration.

**Comment by aleksamagicka** [comment]
Aquasuite works correctly and updates fine for the aquaero. I think I've figured out what's happening. For reference, here's how Aquasuite looks for me on an Aquaero 6:

Aquasuite for Aquaero 6

And here's the excerpt from the manual that I've found (page 29, section 13.2):

- In “power controlled” mode, an output value from a controller will be linearly converted to the range between minimum and maximum power of the fan output and the corresponding voltage will be set.

- In “speed controlled” mode, an output value from a controller will be linearly converted to the range between configured minimum and maximum speed of the fan output. The aquaero will adjust the output power (voltage) autonomously within
the configured range of minimum and maximum power to maintain this fan speed.

- In “PWM controlled” mode, an output value from a controller will be directly set as PWM duty cycle within the limits of minimum and maximum power setting. Fan output voltage will be permanently set to 12 V. If the output value is 0 %, the supply voltage is switched off to stop the fan.

The "power controlled" mode controls the fan using its voltage, and the PWM value we're setting in the driver is used for scaling it - it's for DC fans. The "PWM controlled" mode is for PWM fans, and it correctly sets the voltage to 12V. We'll definitely need to differentiate between those, as my fan (it's a black Noctua A12x15) started clicking in the power controlled mode, as it should only receive stable ~12V.

Secondly, the reason why the fan wasn't starting below 100 PWM is the minimum RPM setting, which was set to 450, which coincides with how this fan treats PWM 100. Setting that to 0 made it spin normally.

In the sysfs docs for hwmon I found this:

```
pwm[1-*]_mode	0: DC mode (direct current)
		1: PWM mode (pulse-width modulation)
		RW
```

That should be OK for differentiating between "power controlled" and "PWM controlled" as the fundamental type of fan that's being controlled. Then, in both modes we can respectively set the minimum RPM (and power?) to 0 to let it run at all speeds. That leaves us with the problem of setting the maximum RPM, or letting the user set it somehow. Maybe that's best done through debugfs, what do you think? (Ignoring the "speed controlled" mode for now.)

**Review by aleksamagicka — APPROVED** [review]

**Comment by leoratte** [comment]
I added the option to switch between dc and pwm. On aquaero 5 pwm is only supported for fan4 as the fans 1-3 are dc only, but this is not yet implemented in the driver.
For controlling min and max rpm sysfs has fan[1-*]_min and fan[1-*]_max. I'll look into implementing them.

**Comment by aleksamagicka** [comment]
Ah, I see them now, thanks. I was too focused on the RPM section.

If it turns out we can't differentiate between aquaero 5 and 6 for the DC only fans, we'll just note it in the docs.

File: aquacomputer_d5next.c [hunk 3]
```diff
@@ -670,8 +731,25 @@ static int aqc_read(struct device *dev, enum hwmon_sensor_types type, u32 attr,
...
+					priv->fan_ctrl_offsets[channel] +
+					AQUAERO_FAN_CTRL_MODE_OFFSET, val, 8);
+			if (ret < 0)
+				return ret;
+
+			switch (*val) {
+			case 0: /* DC mode */
+				break;
+			case 2: /* PWM mode */
+				*val = 1;
```
Comment (aleksamagicka) [inline aquacomputer_d5next.c:746]: Missing a break here.

**Review by aleksamagicka — CHANGES_REQUESTED** [review]
Works great, thanks! Just one comment here, and then I'll merge.

**Comment by leoratte** [comment]
Should be ready to merge now.

**Review by aleksamagicka — APPROVED** [review]

**Comment by aleksamagicka** [comment]
Great, thanks!

**Pull request merged**